src/app/OperationalDeviceProxy.cpp

```diff
@@ -285,31 +294,44 @@ void OperationalDeviceProxy::DequeueConnectionCallbacks(CHIP_ERROR error)
 
 void OperationalDeviceProxy::OnSessionEstablishmentError(CHIP_ERROR error)
 {
-    VerifyOrReturn(mState != State::Uninitialized && mState != State::NeedsAddress,
+    VerifyOrReturn(mState == State::Connecting || mState == State::Recovering,
                    ChipLogError(Controller, "HandleCASEConnectionFailure was called while the device was not initialized"));
 
-    //
-    // We don't need to reset the state all the way back to NeedsAddress since all that transpired
-    // was just CASE connection failure. So let's re-use the cached address to re-do CASE again
-    // if need-be.
-    //
-    MoveToState(State::HasAddress);
+    if (mState == State::Connecting)
+    {
+        //
+        // We don't need to reset the state all the way back to NeedsAddress since all that transpired
+        // was just CASE connection failure. So let's re-use the cached address to re-do CASE again
+        // if need-be.
+        //
```

Comment:
Ignoring the multicast traffic injected for mDNS and only considering the two peers involved in the session, CASE is the expensive part of the operation and address resolution is relatively cheap.  Does it really make sense to undertake a potentially lengthy sequence of retries with a potentially stale address?